### PR TITLE
Fix Duplicate workflow reminder execution in Scheduler mode

### PR DESCRIPTION
**Pull request by JoshVanL**
Workflow reminders must have a period set of repeating forever. This is what fundamentally ensures the durability of Workflows.

PR updates all workflow reminders to have an infinite period.

Workflow reminders repeat execution every 1 minute. In some cases, such as during extremely high load, a workflow reminder may execute whilst a previous execution is still in progress. This leads to situations such as double activity invocation. This patch adds a duplicate internal reminder detector, which silently returns if a duplicate reminder is detected.

Removes the background contexts to Scheduler APIs. The Scheduler API server must respect inbound client contexts.

Removes unused `--placement` flag from the Scheduler.

### CODE REVIEW

**Comment by cicoyle** [comment]
I've confirmed the db is clean after several thousand workflow iterations and the reminders are not triggered indefinitely even while removing the Repeat from 2, should be good with this diff 👍🏻

**Review by cicoyle — DISMISSED** [review]

File: pkg/scheduler/server/api.go [hunk 1]
```diff
@@ -57,9 +57,7 @@ func (s *Server) ScheduleJob(ctx context.Context, req *schedulerv1pb.ScheduleJob
 		Payload:  req.GetJob().GetData(),
 	}
 
-	reqCtx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	err = s.cron.Add(reqCtx, jobName, job)
+	err = s.cron.Add(ctx, jobName, job)
```
Comment (artursouza) [inline pkg/scheduler/server/api.go:60]: This makes sense to me.

File: pkg/runtime/scheduler/streamer.go [hunk 2]
```diff
@@ -110,23 +109,6 @@ func (s *streamer) handleJob(ctx context.Context, job *schedulerv1pb.WatchJobsRe
...
-			// this err is expected if the reminder was triggered once already, the next time it goes to get
-			// triggered by scheduler it sees that it's been canceled and does not invoke the 2nd time. This
-			// more relevant for workflows which currently has a repeat set to 2 since setting it to 1 caused
-			// issues. This will be updated in the future releases, but for now we will see this err. To not
-			// spam users only log if the error is not reminder canceled because this is expected for now.
-			if errors.Is(err, actors.ErrReminderCanceled) {
-				return
-			}
-
-			// If the actor was hosted on another instance, the error will be a gRPC status error,
```
Comment (artursouza) [inline pkg/runtime/scheduler/streamer.go:122]: This was the change that Elena did this morning. Have you validated that the spam of error logs is not present after removing this?

File: pkg/actors/actors.go [hunk 3]
```diff
@@ -1000,6 +995,10 @@ func (a *actorsRuntime) drainRebalancedActors() {
 	// visit all currently active actors.
 	var wg sync.WaitGroup
 
+	a.lock.Lock()
+	a.internalReminderInProgress = make(map[string]struct{})
```
Comment (artursouza) [inline pkg/actors/actors.go:999]: Why is this needed? Don't we have single thread guarantees anymore when using scheduler for reminders?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: cmd/scheduler/options/options.go [hunk 4]
```diff
@@ -73,7 +73,6 @@ func New(origArgs []string) (*Options, error) {
 	fs.StringVar(&opts.TrustDomain, "trust-domain", "localhost", "Trust domain for the Dapr control plane")
 	fs.StringVar(&opts.taFile, "trust-anchors-file", securityConsts.ControlPlaneDefaultTrustAnchorsPath, "Filepath to the trust anchors for the Dapr control plane")
 	fs.StringVar(&opts.SentryAddress, "sentry-address", fmt.Sprintf("dapr-sentry.%s.svc:443", security.CurrentNamespace()), "Address of the Sentry service")
-	fs.StringVar(&opts.PlacementAddress, "placement-address", "", "Addresses for Dapr Actor Placement service")
```
Comment (artursouza) [inline cmd/scheduler/options/options.go:76]: How we missed that? Thanks.

File: pkg/actors/actors.go [hunk 5]
```diff
@@ -116,8 +114,6 @@ type Actors interface {
 	DeleteTimer(ctx context.Context, req *DeleteTimerRequest) error
 	// ExecuteLocalOrRemoteActorReminder executes a reminder on a local or remote actor.
 	ExecuteLocalOrRemoteActorReminder(ctx context.Context, reminder *CreateReminderRequest) error
-	// UsingScheduledReminders returns a bool indicating if the actor implementation is using the scheduler service
-	UsingSchedulerReminders() bool
```
Comment (artursouza) [inline pkg/actors/actors.go:120]: I am happy that we don't need this hack anymore.

File: pkg/actors/actors.go [hunk 6]
```diff
@@ -1103,15 +1102,35 @@ func (a *actorsRuntime) doExecuteReminderOrTimerOnInternalActor(ctx context.Cont
...
+		key := reminder.Key()
+
+		log.Debugf("Executing reminder for internal actor '%s'", key)
+
+		a.lock.Lock()
+		if _, ok := a.internalReminderInProgress[key]; ok {
+			a.lock.Unlock()
+			// We don't need to return cancel here as the first invocation will
+			// delete the reminder.
+			log.Debugf("Duplicate concurrent reminder invocation detected for '%s', likely due to long processing time. Ignoring in favour of the active invocation", key)
```
Comment (artursouza) [inline pkg/actors/actors.go:1114]: This scares me.

File: pkg/runtime/scheduler/streamer.go (hunk 2 above)
Comment (elena-kolevska) [reply pkg/runtime/scheduler/streamer.go:122]: I think we're not going to get this errors anymore, because we're deleting the reminder.

Comment (elena-kolevska) [reply pkg/runtime/scheduler/streamer.go:122]: I think it's because we're now deleting the reminders when completed and returning nil instead of the original ErrReminderCancelled error

File: pkg/actors/actors.go (hunk 3 above)
Comment (JoshVanL) [reply pkg/actors/actors.go:999]: If an internal (workflow) actor reminder execution takes longer than 1m to process, then a duplicate reminder will be executed. To prevent duplication, we need to track what internal actor reminders are in progress to check against.

Existing application reminder guarantees should remain the same

File: pkg/actors/actors.go (hunk 6 above)
Comment (JoshVanL) [reply pkg/actors/actors.go:1114]: We could gate this map check behind the scheduler actor reminder feature gate?

> Is there a bug in how scheduler integrates into reminders that causes this abstraction leak? Firing reminders always had single thread guarantees in Dapr.

> Is there a bug in how scheduler integrates into reminders that causes this abstraction leak? Firing reminders always had single thread guarantees in Dapr.

We are only dealing with internal actor reminders here, which as generally implemented today, are a bit of a hack anyway w.r.t. how actors works generally in application code.

File: pkg/actors/actors.go (hunk 3 above)
Comment (artursouza) [reply pkg/actors/actors.go:999]: As per our call, this code will "ensure" that there is not a memory leak. This code is executed for all actors (internal and external) but is not populating the map. I am fine with this for 1.14 since we plan to revisit internal actors' implementation in 1.15.

File: pkg/runtime/scheduler/streamer.go (hunk 2 above)
Comment (artursouza) [reply pkg/runtime/scheduler/streamer.go]: We are bringing it back as there was no manual verification that this is not needed anymore. Ideally, we should have automation but, at this stage, we need to be conservative.

**Review by artursouza — DISMISSED** [review]

**Review by artursouza — APPROVED** [review]

**Comment by cicoyle** [comment]
LGTM. Just ran my local test and saw the duplicated triggering since the run activity took longer than a minute and saw the log about ignoring it since one is still active, then deleting it. Post workflow run of 4k workflow iterations my db was clean still 👍🏻 

```
scheduler-0-1      | time="2024-08-13T13:49:41.788503301Z" level=debug msg="Triggering job: actorreminder||default||dapr.internal.default.workflow-a.activity||2953-1::0::1||run-activity" instance=bcbc20ef8d3b scope=dapr.scheduler.server type=log ver=edge
scheduler-0-1      | time="2024-08-13T13:50:41.004162967Z" level=debug msg="Triggering job: actorreminder||default||dapr.internal.default.workflow-a.activity||2953-1::0::1||run-activity" instance=bcbc20ef8d3b scope=dapr.scheduler.server type=log ver=edge

workflow-dapr-a-1  | time="2024-08-13T13:50:51.394679014Z" level=debug msg="Executing reminder for internal actor 'dapr.internal.default.workflow-a.activity||2953-1::0::1||run-activity'" app_id=workflow-a instance=e455363315f3 scope=dapr.runtime.actor type=log ver=edge
workflow-dapr-a-1  | time="2024-08-13T13:50:51.394693722Z" level=debug msg="Duplicate concurrent reminder invocation detected for 'dapr.internal.default.workflow-a.activity||2953-1::0::1||run-activity', likely due to long processing time. Ignoring in favour of the active invocation" app_id=workflow-a instance=e455363315f3 scope=dapr.runtime.actor type=log ver=edge
workflow-dapr-a-1  | time="2024-08-13T13:50:51.394715305Z" level=debug msg="Deleting reminder which was cancelled: dapr.internal.default.workflow-a.activity||2953-1::0::1||run-activity" app_id=workflow-a instance=e455363315f3 scope=dapr.runtime.actor type=log ver=edge
```

**Review by cicoyle — APPROVED** [review]

**Review by mikeee — APPROVED** [review]
